### PR TITLE
Add Prometheus shape for metrics

### PR DESCRIPTION
**Pull request by mavam**
## 🔍 Problem

- `metrics` only emits raw `tenzir.metrics.*` event schemas.
- Prometheus export paths need a canonical metric/value shape with labels, units, and metric type metadata.

## 🛠️ Solution

- Add `shape="raw" | "prometheus"` to the describer-backed/neo `metrics` path, keeping `raw` as the default.
- Reshape raw metric slices centrally into `{metric, value, timestamp, labels, type, unit}` records in Prometheus mode.
- Keep raw-mode filter pushdown unchanged, but defer filters in Prometheus mode so shaped fields such as `metric` and `labels.*` exist before filtering.
- Leave the legacy `metrics` parser/factory path unchanged to keep this PR scoped to the neo executor implementation.
- Cover the shaper edge cases in unit tests and add an operator smoke test for post-shape filtering.

## 💬 Review

- Focus on the shaper semantics: recursive flattening, list-of-record handling, label extraction, duration conversion, and descriptor-backed counter/unit decisions.
- The legacy metrics entrypoints intentionally do not parse `shape`; support there was removed after review to keep the change minimal and neo-only.
- Verified locally with `scripts/build.sh tenzir-unit-test`, `scripts/build.sh tenzir`, the focused shaper unit suite, and `uvx tenzir-test --root test --match node/operator_profile_metrics`.

<sub>
📚 Docs PR: tenzir/docs#332<br>
✅ Closes TNZ-428
</sub>

### CODE REVIEW

File: libtenzir/builtins/operators/export.cpp [hunk 1]
```diff
@@ -187,6 +212,17 @@ class Export final : public Operator<void, table_slice> {
...
     }
+    if (uses_prometheus_shape(args_)) {
+      for (auto&& output : detail::shape_metrics_for_prometheus(*expected)) {
+        if (remainder_) {
+          output = filter2(output, *remainder_, ctx, false);
+        }
+        if (output.rows() > 0) {
+          co_await push(std::move(output));
+        }
+      }
```
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp]: Fixed in 32f39d2a170: Prometheus-shaped metrics now increment `read_events_counter_` with the number of shaped rows emitted after post-shape filtering, matching the raw metrics path. Verified locally with `scripts/build.sh tenzir-unit-test`, `scripts/build.sh tenzir`, the focused shaper unit suite, and `uvx tenzir-test --root test --match node/operator_profile_metrics`.

File: libtenzir/builtins/operators/export.cpp [hunk 2]
```diff
@@ -535,6 +573,8 @@ class metrics_plugin final : public virtual operator_parser_plugin,
     auto name = d.positional("name", &ExportArgs::metrics_name);
     d.named("live", &ExportArgs::live);
     d.named("retro", &ExportArgs::retro);
+    auto shape
+      = d.named_optional("shape", &ExportArgs::shape, "raw|prometheus");
```
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp:619]: Addressed by scope clarification rather than code: we intentionally removed `shape` from the legacy metrics parser/factory path to keep TNZ-428 limited to the describer-backed/neo executor implementation. The PR body now calls this out explicitly so the advertised behavior does not imply legacy support.

[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp:619]: Addressed by scope clarification rather than code: we intentionally removed `shape` from the legacy metrics parser/factory path to keep TNZ-428 limited to the describer-backed/neo executor implementation. The PR body now calls this out explicitly so the advertised behavior does not imply legacy support.

File: libtenzir/src/detail/prometheus_metric_shaper.cpp [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  return sanitize_identifier(input, is_label_name_char, false);
+}
+
+auto is_dimension_field(std::string_view field) -> bool {
+  static constexpr auto fields = std::array{
+    "pipeline_id"sv, "operator_id"sv, "name"sv,   "path"sv,        "schema"sv,
+    "schema_id"sv,   "handle"sv,      "method"sv, "status_code"sv,
+  };
+  return std::ranges::find(fields, field) != fields.end();
+}
```
Comment (jachris) [inline libtenzir/src/detail/prometheus_metric_shaper.cpp]: The one big issue I see here is that we have a lot of high-cardinality labels in there. The usual advice is to keep the number of entries for each label low. While you can still argue for `pipeline_id` (although this can already be 100+), things like `schema_id` can change way to frequently.

Comment (mavam) [reply libtenzir/src/detail/prometheus_metric_shaper.cpp]: I agree, and I think the deeper issue is that the shaper currently treats raw Tenzir metric dimensions as Prometheus labels via a global field-name allowlist. That is the wrong abstraction boundary: these fields are cheap query dimensions in Tenzir, but in Prometheus they define time-series identity.

I would not address this by only deleting `schema_id`. A better fix is to make label extraction schema/metric-specific and default-deny:

- no field becomes a Prometheus label unless the metric family explicitly allows it
- drop `schema_id` entirely from the Prometheus shape
- keep only defensible low-cardinality labels, e.g. `method`/`status_code` for API metrics, `path` where it is a bounded canonical path, and `pipeline_id`/`operator_id`/`name` where per-pipeline/operator profiling is the point of the metric
- for import/export/ingest/publish/subscribe schema metrics, either keep only `schema` or aggregate across dropped dimensions so we do not produce duplicate Prometheus series after removing labels

So I’ll rework this as a curated label policy rather than a generic dimension-field promotion rule, and update the docs accordingly.

Comment (mavam) [reply libtenzir/src/detail/prometheus_metric_shaper.cpp]: Implemented in 81f03af1133: label extraction is now metric-family specific and default-deny, `schema_id`/request IDs/native handles/operator metadata are omitted from Prometheus labels, and samples that collapse after omitted metadata are summed. Added regression coverage for schema ID dropping/aggregation and operator metadata, plus docs for the curated label set.

File: libtenzir/builtins/operators/export.cpp [hunk 4]
```diff
@@ -100,23 +131,17 @@ class Export final : public Operator<void, table_slice> {
...
       },
     });
-    for (const auto& filter : args_.filter) {
-      auto [legacy, remainder] = split_legacy_expression(filter);
-      if (legacy != trivially_true_expression()) {
-        legacy_clauses.push_back(std::move(legacy));
+    if (uses_prometheus_shape(args_)) {
+      for (const auto& filter : args_.filter) {
+        add_remainder(remainder_, filter);
       }
```
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp]: Intentionally not changing this: `shape="prometheus"` is only supported on the describer-backed/neo path in this PR, and filters in that mode are post-shape TQL2 filters. Preserving TQL1-style raw/meta pushdown here would add legacy-path behavior we explicitly want to keep out of scope.

File: libtenzir/builtins/operators/export.cpp [hunk 5]
```diff
@@ -87,6 +141,9 @@ class Export final : public Operator<void, table_slice> {
       },
       MetricsDirection::read, MetricsVisibility::internal_,
       MetricsUnit::events);
+    if (uses_prometheus_shape(args_)) {
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/export.cpp]: `uses_prometheus_shape` should be a boolean member of the operator, initialized exactly once on construction.

File: libtenzir/builtins/operators/export.cpp [hunk 6]
```diff
@@ -187,6 +238,24 @@ class Export final : public Operator<void, table_slice> {
       done_ = true;
       co_return;
     }
+    if (uses_prometheus_shape(args_)) {
+      auto shaper
+        = prometheus_shapers_.find(std::string{expected->schema().name()});
+      if (shaper == prometheus_shapers_.end()) {
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/export.cpp:244]: We probably still want a warning here, since we are omitting output.

File: libtenzir/builtins/operators/export.cpp [hunk 7]
```diff
@@ -67,8 +72,57 @@ struct ExportArgs {
...
+  std::string shape = "raw";
 };
 
+auto uses_prometheus_shape(ExportArgs const& args) -> bool {
+  return args.special_filter == export_special_filter::metrics
+         and from_string<metrics_shape>(args.shape).value_or(metrics_shape::raw)
+               == metrics_shape::prometheus;
+}
+
+using prometheus_shaper_map
```
Comment (IyeOnline) [inline libtenzir/builtins/operators/export.cpp:84]: Use libtenzir/include/tenzir/detail/heterogeneous_string_hash.hpp here, then you wont have to explicitly materiaize `std::string`s on lookup

File: libtenzir/src/detail/prometheus_metric_shaper.cpp [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+  if (ascii_isdigit(first) or (not allow_colon and result.front() == ':')) {
+    result.insert(result.begin(), '_');
+  }
+  return result;
+}
+
+auto prometheus_role_of(type const& field_type) -> prometheus_role {
+  auto const role = field_type.attribute(metrics::prometheus_role_attribute);
+  if (not role) {
+    return prometheus_role::none;
```
Comment (IyeOnline) [inline libtenzir/src/detail/prometheus_metric_shaper.cpp:149]: I believe the `prometheus_role::none` value is not handled correctly everywhere. A lot of code branches based on ignore or "not X", but fields that dont have a role at all presumably should be treated as-if ignored.

File: libtenzir/src/detail/prometheus_metric_shaper.cpp [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+  auto labels_record = metric.field("labels").record();
+  for (auto const& [key, label_value] : sample.metric_labels) {
+    labels_record.field(key, label_value);
+  }
+  metric.field("type", sample.type);
+  metric.field("unit", sample.unit);
+}
+
+auto flatten_value(std::vector<metric_sample>& samples, std::string_view source,
+                   std::vector<std::string>& path, data_view3 value,
```
Comment (IyeOnline) [inline libtenzir/src/detail/prometheus_metric_shaper.cpp]: There is value_path.hpp, which is a superior version of manually doing this path handling.

**Review by IyeOnline — COMMENTED** [review]
I _think_ it should be possible to do this entire transformation in a columnar fashion at the arrow level.

The only complication are lists, which complicate unpacking here, given you need to avoid factoring. _However_ the only actual use-cases for this are very detailed by actor CAF and memory metrics; both of which I highly doubt are something users care about.

You should be able to just build the label/unit columns and then combine them with the data column. No builder needed at all.

// edit: I think it may even be possible to do this for the list columns by just taking their value column, and building a new timestamp column to go alongside it.

Maybe look into that, as it should also simplify the code itself (besides making it more performant). This is also only a hunch, so take it with a grain of salt.

---

I still left a few other comments before I realized this may be possible

File: libtenzir/builtins/operators/export.cpp (hunk 5 above)
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp]: Fixed in aa403e606f0: `uses_prometheus_shape_` is now initialized once in the operator constructor and reused by `start`/`process_task`.

File: libtenzir/builtins/operators/export.cpp (hunk 6 above)
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp:244]: Fixed in aa403e606f0: unsupported Prometheus-shape metric schemas now emit a warning once before the slice is omitted.

File: libtenzir/builtins/operators/export.cpp (hunk 7 above)
Comment (mavam) [reply libtenzir/builtins/operators/export.cpp:84]: Fixed in aa403e606f0: the shaper map now uses `heterogeneous_string_hashmap`, so schema-name lookup no longer materializes a temporary `std::string`.

File: libtenzir/src/detail/prometheus_metric_shaper.cpp (hunk 8 above)
Comment (mavam) [reply libtenzir/src/detail/prometheus_metric_shaper.cpp:149]: Fixed in aa403e606f0: unannotated leaf fields are ignored, while unannotated record/list containers are still traversed so annotated descendants can be shaped.

File: libtenzir/src/detail/prometheus_metric_shaper.cpp (hunk 9 above)
Comment (mavam) [reply libtenzir/src/detail/prometheus_metric_shaper.cpp]: Fixed in aa403e606f0: metric-name path tracking now uses `value_path` instead of a manual string vector.
